test/box-luatest/gh_6150_memtx_tx_memory_monitoring_test.lua

```diff
@@ -0,0 +1,200 @@
+local server = require('test.luatest_helpers.server')
+local t = require('luatest')
+
+local g = t.group()
+
+local current_stat = {}
+
+local function table_apply_change(table, related_changes)
+    for k, v in pairs(related_changes) do
+        if type(v) ~= 'table' then
+            table[k] = table[k] + v
+        else
+            table_apply_change(table[k], v)
+        end
+    end
+end
+
+local function table_values_are_zeros(table)
+    for _, v in pairs(table) do
+        if type(v) ~= 'table' then
+            if v ~= 0 then
+                return false
+            end
+        else
+            if not table_values_are_zeros(v) then
+                return false
+            end
+        end
+    end
+    return true
+end
+
+local function tx_gc(server, steps, related_changes)
+    server:eval('box.internal.memtx_tx_gc(' .. steps .. ')')
+    if related_changes then
+        table_apply_change(current_stat, related_changes)
+    end
+    assert(table.equals(current_stat, server:eval('return box.stat.memtx.mvcc()')))
+end
+
+local function tx_step(server, txn_name, op, related_changes)
+    server:eval(txn_name ..  '("' .. op .. '")')
+    if related_changes then
+        table_apply_change(current_stat, related_changes)
+    end
+    assert(table.equals(current_stat, server:eval('return box.stat.memtx.mvcc()')))
+end
+
+g.before_each(function()
+    g.server = server:new{
+        alias   = 'default',
+        box_cfg = {memtx_use_mvcc_engine = true}
+    }
+    g.server:start()
+
+    g.server:eval('txn_proxy = require("test.box.lua.txn_proxy")')
+    g.server:eval('s = box.schema.space.create("test")')
+    g.server:eval('s:create_index("pk")')
+    g.server:eval('box.internal.memtx_tx_gc(100)')
+    -- CREATING CURRENT STAT
+    current_stat = g.server:eval('return box.stat.memtx.mvcc()')
+    -- Check if txm use no memory
+    assert(table_values_are_zeros(current_stat))
+end)
+
+g.after_each(function()
+    -- Check if there is no memory occupied by txm
+    assert(table_values_are_zeros(current_stat))
+    g.server:drop()
+end)
+
+g.test_memtx_tx_mvcc_mem = function()
+    g.server:eval('tx1 = txn_proxy.new()')
+    g.server:eval('tx2 = txn_proxy.new()')
+    g.server:eval('tx1:begin()')
+    g.server:eval('tx2:begin()')
+    local diff = {
+        ["txn"] = {
+            ["statements"] = {
+                ["avg"] = 68,
+                ["total"] = 136,
+                ["max"] = 136
+            },
+            ["system data"] = {
+                ["total"] = 147,
+                ["max"] = 147,
+                ["avg"] = 73
+            }
+        },
+        ["memtx mvcc"] = {
+            ["tuple stories"] = {
+                ["directly used by txn"] = {
+                    ["stories total"] = 152,
+                    ["stories count"] = 1,
+                }
+            }
+        }
+    }
```

Comment:
I really don't like that your test relies on allocation sizes: at least, it might be hard to maintain.

I think getting the allocation size delta of one story from box.stat.memtx.mvcc() should suffice.

Reply:
The idea of the test was that at the slightest change in memory consumption the test would fail, and we noticed that.